### ISSUE TITLE
[HOLD for payment 2024-01-22] [$1000] Desktop - FAB - Modal for 'Send Message' freezes when opened

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#28916]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open the testing desktop app
2. Click on the FAB button
3. Click on Send message
4. Repeat 2-3 times

## Expected Result:
Modal opens without any lags or freezes

## Actual Result:
Modal freezes when it is opened / sometimes the sliding motion is neglected and it appears immediately without any sliding effect

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** 1.3.78.0

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/78819774/39206801-5320-47fb-9d9c-2347b8476204

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01a81bbd7ec35b9394</li>
        <li>Upwork Job ID: 1709988778512769024</li>
        <li>Last Price Increase: 2023-11-06</li>
    </ul>
</details>

**Comment by AmjedNazzal** [comment]
Is this only on staging with beta? I can't repo on dev with beta on

**Comment by studentofcoding** [comment]
# Proposal

## Please re-state the problem that we are trying to solve in this issue

On Desktop - FAB - Modal for 'Send Message' freezes when opened

## What is the root cause of that problem?

The root cause of the problem is due to heavy rendering on [src/pages/NewChatSelectorPage.js](https://github.com/Expensify/App/blob/619aee628a82adb85327dfb95f5f1280c6970a95/src/pages/NewChatSelectorPage.js) when the modal is opened that causing animation delay in the rendering process.

## What changes do you think we should make in order to solve the problem?

We should optimize the rendering of the components in the NewChatSelectorPage modal. via lazy loading `NewChatPage` to load the component only when needed. 

Here below is the example code, with the result

```
import React, {lazy} from 'react';
//.. Other codes

// Lazy load NewChatPage, so the animation is smoother and load the data after Modal is opened
const NewChatPage = lazy(() => import('./NewChatPage'));

//.. Other codes

function NewChatSelectorPage(props) {
    return (
        <ScreenWrapper
            shouldEnableKeyboardAvoidingView={false}
            includeSafeAreaPaddingBottom={false}
            shouldEnableMaxHeight
            testID={NewChatSelectorPage.displayName}
        >
            <HeaderWithBackButton title={props.translate('sidebarScreen.fabNewChat')} />
            {Permissions.canUsePolicyRooms(props.betas) ? (
                <OnyxTabNavigator
                    id={CONST.TAB.NEW_CHAT_TAB_ID}
                    tabBar={({state, navigation, position}) => (
                        <TabSelector
                            state={state}
                            navigation={navigation}
                            position={position}
                        />
                    )}
                >
                    <TopTab.Screen
                        name={CONST.TAB.NEW_CHAT}
                        component={() => (
                            //.. Use Suspense to make sure that the component is loaded
                            //.. we can use another component as the fallback as well
                            <Suspense fallback={null}>
                                <NewChatPage />
                            </Suspense>
                        )}
                    />
                    <TopTab.Screen
                        name={CONST.TAB.NEW_ROOM}
                        component={WorkspaceNewRoomPage}
                    />
                </OnyxTabNavigator>
            ) : (
                //.. Use Suspense to make sure that the component is loaded
                //.. we can use another component as the fallback as well
                <Suspense fallback={null}>
                    <NewChatPage />
                </Suspense>
            )}
        </ScreenWrapper>
    );
}

NewChatSelectorPage.propTypes = propTypes;
NewChatSelectorPage.defaultProps = defaultProps;
NewChatSelectorPage.displayName = 'NewChatPage';

export default compose(
    withLocalize,
    withWindowDimensions,
    withOnyx({
        betas: {key: ONYXKEYS.BETAS},
    }),
)(NewChatSelectorPage);
```

## Result

https://github.com/Expensify/App/assets/20473526/c39afd56-74e8-4da1-97da-640a9f34cd30

## What alternative solutions did you explore? (Optional)
NA

@eVoloshchak @lschurr

**Comment by eVoloshchak** [comment]
> The root cause of the problem is due to heavy rendering on [src/pages/NewChatSelectorPage.js](https://github.com/Expensify/App/blob/619aee628a82adb85327dfb95f5f1280c6970a95/src/pages/NewChatSelectorPage.js)

@studentofcoding, lazy loading the component would fix the symptom of the problem, but not the root cause.
What is particularly heavy about this page? What's causing those heavy renders?

**Comment by studentofcoding** [comment]
> > The root cause of the problem is due to heavy rendering on [src/pages/NewChatSelectorPage.js](https://github.com/Expensify/App/blob/619aee628a82adb85327dfb95f5f1280c6970a95/src/pages/NewChatSelectorPage.js)
> 
> 
> 
> @studentofcoding, lazy loading the component would fix the symptom of the problem, but not the root cause.
> 
> What is particularly heavy about this page? What's causing those heavy renders?

It's on Search List that get fetch & rendered everytime we open the modal @eVoloshchak

**Comment by eVoloshchak** [comment]
But we don't have the same problem when opening the search page (CMD+K). Is it due to skeleton UI?

**Comment by studentofcoding** [comment]
> But we don't have the same problem when opening the search page (CMD+K). Is it due to skeleton UI?

Yes it's @eVoloshchak

**Comment by lschurr** [comment]
Any update here @eVoloshchak?

**Comment by eVoloshchak** [comment]
This is awaiting more proposals
I think we need to dig deeper to figure out and fix the actual root cause for this

**Comment by eVoloshchak** [comment]
Not overdue, this is awaiting more proposals
I think we need to dig deeper to figure out and fix the actual root cause for this

**Comment by akamefi202** [comment]
I'm not able reproduce this bug.

**Comment by eVoloshchak** [comment]
@akamefi202, it's still reproducible for me 

https://github.com/Expensify/App/assets/9059945/bf65dbc8-17fc-4dce-8314-b7cc2e01b7fd

It's present only on Desktop and only for accounts with `policyRooms` beta (or just override `canUsePolicyRooms` to return true

**Comment by lschurr** [comment]
Are we still waiting for proposals @eVoloshchak?

**Comment by eVoloshchak** [comment]
@lschurr, yes, this is reproducible, still waiting for proposals

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by eVoloshchak** [comment]
This is still reproducible for me. To test this your account has to be added to the `policy rooms` beta

https://github.com/Expensify/App/assets/9059945/69cd7b00-94a8-4e96-8bcc-1ed6c294b87b

**Comment by eVoloshchak** [comment]
Not overdue, still awaiting proposals

**Comment by Julesssss** [comment]
Still awaiting proposals. @lschurr maybe we can double the bounty?

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Second week)

**Comment by lschurr** [comment]
Alright, I've updated the bounty on this one.

**Comment by eVoloshchak** [comment]
Still reproducible with high traffic account and canUseAllBetas=true 

https://github.com/Expensify/App/assets/9059945/c90d4993-b001-4362-968d-eca94ee8f2ed

**Comment by ikevin127** [comment]
While most likely this proposal will not meet the requirements, I'll give it a try anyways because I spent some time / had some fun debugging this and  maybe some of my observations might be helpful to the next person attempting to solve it.

# Proposal

## Please re-state the problem that we are trying to solve in this issue

Modal freezes when it is opened / sometimes the sliding motion is neglected and it appears immediately without any sliding effect.

## What is the root cause of that problem?

The issue is caused by the implementation of `SectionList` from [this](https://github.com/Expensify/App/blob/9b534633dcc1563d3410d6371ee1537de5e9a1f2/src/components/OptionsList/BaseOptionsList.js#L38) file.

When the right hand navigation modal is opened the component containing this list is also being rendered which causes the freeze due to multiple re-renders of the list loading new options (items)  and expensive calculations of the layout every time the list items are rendered but especially the very first time the modal is opened.

**Debugging observations:**
- the first time the modal is animating in is the worst in terms of performance, it gets better subsequently
- commenting out the avatar images does improve performance of `SectionList`
- the freeze happens on web too, using a high traffic account and canUseAllBetas=true
- opening the right hand havigator only with the `OptionsListSkeletonView` by itself causes performance issues, it's not just the `SectionList` which leads one to think that the problem comes from `BaseOptionsList` component which contains all the heavy computations related to the `SectionList`, including the skeleton loading component
- tweaking `SectionList` props like **initialNumToRender**, **maxToRenderPerBatch** and **windowSize** does not seem to help with this issue

## What changes do you think we should make in order to solve the problem?

My solution does not fix the problem, but it offers an improvement in performance.
My approach was to use `InteractionManager.runAfterInteractions()` with `requestAnimationFrame()` inside it in order to delay the rendering of the `OptionsSelector` [component](https://github.com/Expensify/App/blob/9b534633dcc1563d3410d6371ee1537de5e9a1f2/src/pages/NewChatPage.js#L240C35-L240C35) which includes the search input and `SectionList` like so:

```
    // src/pages/NewChatPage.js
    
    const [hasFinishedInteractions, setHasFinishedInteractions] = useState(false);

    useEffect(() => {
        const runAfterInteractionsTask = InteractionManager.runAfterInteractions(() => {
            requestAnimationFrame(() => {
                setHasFinishedInteractions(true);
            });
        });

        return () => {
            if (runAfterInteractionsTask) {
                runAfterInteractionsTask.cancel();
            }
            setHasFinishedInteractions(false);
        }
    }, []);
    
    ...
    // lower in the component return
    
    {hasFinishedInteractions && (
        <OptionsSelector
    ...
```

This improves the performance slightly by delaying the rendering of the `OptionsSelector` component while the navigation transition animation takes place which causes this performance issue.

Looking closely at the videos then going back / forward with the playback cursor we can observe:
- **after**: the transition-in animation can be observed right from the first opening and also subsequently, not the smoothest but still noticeable compared to before
- **before**: the modal is just appearing from none to opened on the first opening, no transition-in animation, then there's barely any transition-in animation going on subsequently which is also very janky

 ### Videos

<details>
<summary>MacOS: Desktop (after / before)</summary>

- after

https://github.com/Expensify/App/assets/56457735/d5f8a98d-0eed-4e24-98c6-358cb8af52a1

- before

https://github.com/Expensify/App/assets/56457735/47cab1ca-3fd2-476b-8811-02d63cabb876

</details>

**Comment by lschurr** [comment]
@eVoloshchak could you review the proposal?

**Comment by eVoloshchak** [comment]
@ikevin127, a huge thank you for posting your findings here

I don't think the solution you're proposing is the best approach, since
- as you've said yourself, it doesn't fix the actual problem
- it still doesn't look quite right on videos, since there is a brief flash before the list is rendered

However, recently we've put more emphasis on improving the app performance, I think there is a plan to improve the FlatList performance specifically (I'll try to find the actual discussion on Slack and link it here), which might resolve this issue

**Comment by lschurr** [comment]
Any luck on finding the supporting Slack conversation @eVoloshchak?

**Comment by lschurr** [comment]
Any update on this one @Julesssss? Are we holding on this issue?

**Comment by Julesssss** [comment]
It was added to the performance project, but doesn't have a status yet.

@muttmuure do we need to do anything else here?

**Comment by muttmuure** [comment]
Sorry I missed this last week - I will call it out asking for a volunteer when I do the weekly update this week

**Comment by muttmuure** [comment]
Is this to do with FlatList specifically? 

@TMisiukiewicz and @adhorodyski might have some insights into what is going wrong here

**Comment by muttmuure** [comment]
Going to try to get the performance update out tonight

**Comment by muttmuure** [comment]
Update went out - @hurali97 and @kacper-mikolajczak are going to take a look

**Comment by hurali97** [comment]
Hey 👋 

We have some findings from our investigation and we want to share those with you. 

- The main outlier is the execution of the `useEffect`  inside `NewChatPage` even before the animation finishes but since the component is mounted, the effect is called. The ideal way to handle that is to wait for the animation to complete and then execute the function. We have a prop called `onEntryTransitionEnd` on `ScreenWrapper` we can use that and move our `useEffect` body to a new method and pass it to the `onEntryTransitionEnd`. Then we can also leverage another prop `didScreenTransitionEnd` from `ScreenWrapper` and use that to decide to show Skeleton Loader by using `isOptionsDataReady && didScreenTransitionEnd`. Alternatively, we can solve this by using `InteractionManager.runAfterInteractions` to wait until the animations are completed.

The above fixes the issue but during the investigation we also found that there are many react commits when we close the Modal and they amount to more than 130+ in numbers, to fix those we suggest following options:

- The `Button` component re-renders a lot because it's used in `OptionRow` for the `Add To Group` UI element. Inside of `Button` component we have `useIsFocused` and `useActiveElement`. Now, since it's a part of the `renderItem` each time any of those changes, part of our `renderItem` will render, which is `Button` and it's children in this case.

  - First improvement is to move this code which is required for keyboard shortcut to outside of `Button` component and call it something like `KeyboardShortcutComponent` and in there use `useIsFocused` and `useActiveElement` and keyboard shortcut code. This way, we will only re-render `KeyboardShortcutComponent` and not the whole `Button` component.
  
  - Second, we can remove `useIsFocused` in favour of `useFocusEffect` and store that boolean value in a ref. The reason here is we are only using `isFocused` in the `keyboardShortcutCallback` which will take updated value each time the function is called, so we can safely remove the state dependency on `useIsFocused` hook.
  
  - Third, we can refactor the code for `useActiveElement` to `useActiveElementRole` because we are only using the `role` property from it, so there seems no point in storing the whole HTML there. On top of that, since the value from this hook is used to evaluate `config` for `useKeyboardShortcut` hook, so we can refactor the state implementation in `useActiveElement` to ref based.

Lastly, we will also benefit from the improvements in [this](https://github.com/Expensify/App/pull/31354) PR, since it targets the transition improvements for `SearchPage` on slower devices. For example, `BaseOptionsSelector` also underperforms in our case, which is optimised in the linked PR. Once that PR is merged, we can simply sync main in our new PR that we will create.

The results below are from the main branch and from all of the improvements mentioned above ( including some parts of the linked PR )

### Result from main:

https://github.com/Expensify/App/assets/47336142/b7e2d0d5-3199-45ca-bbff-b44a062dd48a

### Result from Improvements:

https://github.com/Expensify/App/assets/47336142/05c5aae9-9918-4546-8f49-9cc7ebe76c89

<br />

**Comment by lschurr** [comment]
What's next on this one @eVoloshchak @muttmuure @Julesssss?

**Comment by Julesssss** [comment]
@lschurr the PR was just assigned to the C+ for review, so nothing from us yet

**Comment by Julesssss** [comment]
Seems like we'll need to fix [this](https://github.com/Expensify/App/pull/33055#issuecomment-1875740252). Did you see the comment, @hurali97?

**Comment by hurali97** [comment]
@Julesssss yeah, I will try to fix that by Monday, since we are off tomorrow 👍

**Comment by shubham1206agra** [comment]
Hello @muttmuure @hurali97 @Julesssss @lschurr 

I have tried to further optimize @hurali97's [proposal](https://github.com/Expensify/App/issues/28916#issuecomment-1853920181) here https://github.com/Expensify/App/pull/34480.

I tried to do 2 main things 
1. Instead of creating 2 new listener for `focusin`, and `focusout` for every use of `useActiveElementRole`, we are going to create context of `activeRole`, and will listen only once for events inside the Provider, which will be supplied to App root component directly. The aim was to remove the redundancy of listeners, as we know ever button will have these listeners, and there are almost no places on App where we don't have a button.
2. Instead of using the above context for every button rendered, we have used `pressOnEnter` prop to conditionally render `KeyboardShortcutComponent`, which will help in less re-render on changing of `activeRole`.

Now we use state for storing `activeRole` instead of ref cause this caused a regression https://github.com/Expensify/App/issues/34474, and restoring to using state directly causing some performance lag.

If you agree with me, @hurali97 can you profile https://github.com/Expensify/App/pull/34480 to make sure there is no performance degradation?

**Comment by Julesssss** [comment]
oh I missed this notification

**Comment by Julesssss** [comment]
> I have tried to further optimize @hurali97's https://github.com/Expensify/App/issues/28916#issuecomment-1853920181 here https://github.com/Expensify/App/pull/34480.

We're going to go ahead with this change on [this](https://github.com/Expensify/App/issues/34474) linked PR

**Comment by muttmuure** [comment]
Not overdue

**Comment by muttmuure** [comment]
Is the problem resolved? Can we go ahead and pay?

**Comment by lschurr** [comment]
This looks correct. @eVoloshchak - you can request payment via newdot.

Closing.

**Comment by situchan** [comment]
There were several regressions. One is not resolved yet.
- https://github.com/Expensify/App/issues/34367
- https://github.com/Expensify/App/issues/34371
- https://github.com/Expensify/App/issues/34474

**Comment by eVoloshchak** [comment]
@hurali97, could you take a look at the comment above?

**Comment by eVoloshchak** [comment]
@hurali97, bump on the above

**Comment by eVoloshchak** [comment]
@hurali97, bump on the above
